File: Lab08Cpp/Room.cpp
#include "Room.hpp"
#include <iostream>

void Room::CheckAllSensors()
{
    std::cout << "Checking sensors in room '" << m_name << "'..." << std::endl;
    for (const auto& s : m_sensors)
    {
        if (!s)
            continue;

        s->ReadValue();
        s->CheckStatus();
    }
}

std::shared_ptr<Room> Room::ShallowClone() const
{
    auto clone = std::make_shared<Room>(*this); // копируем name и вектор указателей
    std::cout << "Room '" << m_name << "' shallow-cloned." << std::endl;
    return clone;
}

std::shared_ptr<Room> Room::DeepClone() const
{
    auto clone = std::make_shared<Room>(m_name);
    for (const auto& s : m_sensors)
    {
        if (s)
        {
            clone->AddSensor(SensorPtr(s->Clone())); // новый Sensor с тем же состоянием
        }
    }
    std::cout << "Room '" << m_name << "' deep-cloned." << std::endl;
    return clone;
}

File: Lab08Cs/Lab06Monitoring/Room.cs
using System;
using System.Collections.Generic;

namespace Lab06Monitoring
{
    /// <summary>
    /// Помещение, содержащее набор датчиков.
    ///
    /// ЛР-6:
    ///  * реализуем поверхностное и глубокое клонирование (через Clone датчиков);
    /// </summary>
    public class Room : ICloneable
    {
        private readonly List<Sensor> _sensors = new();

        public string Name { get; }

        public Room(string name)
        {
            Name = name;
        }

        public void AddSensor(Sensor sensor)
        {
            _sensors.Add(sensor);
        }

        public void CheckAllSensors()
        {
            Console.WriteLine($"Checking sensors in room '{Name}'...");
            foreach (var sensor in _sensors)
            {
                sensor.ReadValue();
                sensor.CheckStatus();
            }
        }

        /// <summary>
        /// Поверхностное клонирование комнаты (датчики те же).
        /// </summary>
        public object Clone()
        {
            var clone = new Room(Name);
            foreach (var s in _sensors)
                clone.AddSensor(s); // те же ссылки на датчики

            Console.WriteLine($"Room '{Name}' shallow-cloned.");
            return clone;
        }

        /// <summary>
        /// Глубокое клонирование комнаты (датчики клонируются).
        /// </summary>
        public Room DeepClone()
        {
            var clone = new Room(Name);
            foreach (var s in _sensors)
            {
                if (s.Clone() is Sensor sensorClone)
                    clone.AddSensor(sensorClone);
            }

            Console.WriteLine($"Room '{Name}' deep-cloned.");
            return clone;
        }


        /// <summary>
        /// Возвращает только чтение списка датчиков в комнате.
        /// Используется в ЛР‑8 для построения контейнера всех датчиков в системе.
        /// </summary>
        public IReadOnlyList<Sensor> GetSensorsReadonly()
        {
            return _sensors.AsReadOnly();
        }
    }
}
